File: 0763-partition-labels/0763-partition-labels.py
class Solution:
  
  # O(n) time and O(1) space where n is the length of the input string
  def partitionLabels(self, s: str) -> List[int]:
    lastIndexes = {}

    for i, char in enumerate(s):
        lastIndexes[char] = i

    currSize = 0
    currLastIndex = 0
    output = []

    for i, char in enumerate(s):
      currSize += 1 
      currLastIndex = max(lastIndexes[char], currLastIndex)
      if i == currLastIndex:
        output.append(currSize)
        currSize = 0

    return output

File: 0763-partition-labels/0763-partition-labels.js
/**
 * @param {string} s
 * @return {number[]}
 */
const partitionLabels = (s) => {
  const lastIndexes = {};

  for (let i = 0; i < s.length; i++) {
    lastIndexes[s[i]] = i;
  }

  const output = [];
  let currSize = 0;
  let currLastIdx = 0;

  for (let i = 0; i < s.length; i++) {
    if (lastIndexes[s[i]] > currLastIdx) {
      currLastIdx = lastIndexes[s[i]];
    }

    currSize++;

    if (i === currLastIdx) {
      output.push(currSize);
      currSize = 0;
    }
  }

  return output;
};